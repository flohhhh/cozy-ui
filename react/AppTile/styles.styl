@require 'settings/breakpoints.styl'


.AppTile-icon
  object-fit contain
  max-width 100%
  width auto

.AppTile-icon--default
  height 100%
  padding 0 .5em

