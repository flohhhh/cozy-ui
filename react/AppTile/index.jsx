import React from 'react'

import AppIcon from '../AppIcon'
import PropTypes from 'prop-types'
import Tile, {
  TileTitle,
  TileSubtitle,
  TileFooter,
  TileIcon,
  TileDescription
} from '../Tile'

import { useI18n } from '../I18n'

import { getCurrentStatusLabel } from './helpers'
import { AppDoctype } from '../proptypes'
import styles from './styles.styl'
let dataset
const getDataset = () => {
  if (dataset) return dataset
  const root = document.querySelector('[role=application]')
  dataset = root && root.dataset
  return dataset
}

const getAppIconProps = () => ({
  domain: getDataset() && getDataset().cozyDomain,
  secure: window.location.protocol === 'https:'
})

export const AppTile = ({
  app,
  name,
  namePrefix,
  onClick,
  showDeveloper,
  IconComponent
}) => {
  const { t } = useI18n()
  const { developer = {} } = app
  const statusToDisplay = getCurrentStatusLabel(app)
  IconComponent = IconComponent || AppIcon
  return (
    <Tile tag="button" type="button" onClick={onClick}>
      <TileIcon>
        <IconComponent
          app={app}
          className={styles['AppTile-icon']}
          {...getAppIconProps()}
        />
      </TileIcon>
      <TileDescription>
        <TileTitle>{namePrefix ? `${namePrefix} ${name}` : name}</TileTitle>
        {developer.name && showDeveloper && (
          <TileSubtitle>{`${t('app_item.by')} ${developer.name}`}</TileSubtitle>
        )}
        {statusToDisplay && (
          <TileFooter>{t(`app_item.${statusToDisplay}`)}</TileFooter>
        )}
      </TileDescription>
    </Tile>
  )
}

AppTile.propTypes = {
  app: AppDoctype,
  name: PropTypes.string.isRequired,
  namePrefix: PropTypes.string,
  onClick: PropTypes.func,
  showDeveloper: PropTypes.bool
}

AppTile.defaultProps = {
  showDeveloper: true
}

export default AppTile
