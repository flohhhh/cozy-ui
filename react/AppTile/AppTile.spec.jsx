'use strict'

/* eslint-env jest */

import React from 'react'
import { shallow } from 'enzyme'

import en from '../AppSections/locales/en.json'
import I18n from '../I18n'

import AppTile from './index'

const appMock = {
  slug: 'test',
  editor: 'cozy',
  namePrefix: 'Cozy',
  developer: {
    name: 'Cozy'
  },
  icon: '<svg></svg>',
  name: 'Test',
  version: '3.0.3',
  onClick: jest.fn()
}

const appMock2 = {
  slug: 'test2',
  editor: '',
  developer: {
    name: 'Naming me'
  },
  icon: '<svg></svg>',
  name: 'Test2',
  version: '3.0.3-beta7483',
  installedAppLink: 'test2.cozy.mock',
  installed: true,
  onClick: jest.fn()
}

const Wrapper = props => {
  return (
    <I18n dictRequire={() => en} lang="en">
      <AppTile {...props} />
    </I18n>
  )
}

describe('AppTile component', () => {
  it('should render correctly an app', () => {
    const component = shallow(
      <Wrapper {...appMock} app={appMock} />
    ).getElement()
    expect(component).toMatchSnapshot()
  })

  it('should render correctly an app without developer property', () => {
    const appMockWithoutDevelopper = Object.assign({}, appMock)
    delete appMockWithoutDevelopper.developer
    const component = shallow(
      <Wrapper {...appMockWithoutDevelopper} app={appMockWithoutDevelopper} />
    ).getElement()
    expect(component).toMatchSnapshot()
  })

  it('should render correctly an app in maintenance', () => {
    const appInMaintenance = Object.assign({}, appMock2, {
      maintenance: { maintenance_options: {} }
    })
    const component = shallow(
      <Wrapper {...appInMaintenance} app={appInMaintenance} />
    ).getElement()
    expect(component).toMatchSnapshot()
  })

  it('should render correctly an app with iconToLoad (icon being fetching)', () => {
    const appIconToLoad = Object.assign({}, appMock2, {
      iconToLoad: true
    })
    const component = shallow(
      <Wrapper {...appIconToLoad} app={appIconToLoad} />
    ).getElement()
    expect(component).toMatchSnapshot()
  })

  it('should render correctly an app with iconToLoad if isMobile', () => {
    const appIconToLoad = Object.assign({}, appMock2, {
      iconToLoad: true
    })
    const component = shallow(
      <Wrapper {...appIconToLoad} app={appIconToLoad} isMobile />
    ).getElement()
    expect(component).toMatchSnapshot()
  })

  it('should render correctly an installed app', () => {
    const component = shallow(
      <Wrapper {...appMock2} app={appMock2} />
    ).getElement()
    expect(component).toMatchSnapshot()
  })

  it('should render correctly an installed app without icon provided', () => {
    const appMockWithoutIcon = Object.assign({}, appMock, { icon: '' })
    const component = shallow(
      <Wrapper {...appMockWithoutIcon} app={appMockWithoutIcon} />
    ).getElement()
    expect(component).toMatchSnapshot()
  })
})
