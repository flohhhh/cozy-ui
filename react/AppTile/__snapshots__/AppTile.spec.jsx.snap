// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppTile component should render correctly an app 1`] = `
<BaseI18n
  defaultLang="en"
  dictRequire={[Function]}
  lang="en"
>
  <AppTile
    app={
      Object {
        "developer": Object {
          "name": "Cozy",
        },
        "editor": "cozy",
        "icon": "<svg></svg>",
        "name": "Test",
        "namePrefix": "Cozy",
        "onClick": [MockFunction],
        "slug": "test",
        "version": "3.0.3",
      }
    }
    developer={
      Object {
        "name": "Cozy",
      }
    }
    editor="cozy"
    icon="<svg></svg>"
    name="Test"
    namePrefix="Cozy"
    onClick={[MockFunction]}
    showDeveloper={true}
    slug="test"
    version="3.0.3"
  />
</BaseI18n>
`;

exports[`AppTile component should render correctly an app in maintenance 1`] = `
<BaseI18n
  defaultLang="en"
  dictRequire={[Function]}
  lang="en"
>
  <AppTile
    app={
      Object {
        "developer": Object {
          "name": "Naming me",
        },
        "editor": "",
        "icon": "<svg></svg>",
        "installed": true,
        "installedAppLink": "test2.cozy.mock",
        "maintenance": Object {
          "maintenance_options": Object {},
        },
        "name": "Test2",
        "onClick": [MockFunction],
        "slug": "test2",
        "version": "3.0.3-beta7483",
      }
    }
    developer={
      Object {
        "name": "Naming me",
      }
    }
    editor=""
    icon="<svg></svg>"
    installed={true}
    installedAppLink="test2.cozy.mock"
    maintenance={
      Object {
        "maintenance_options": Object {},
      }
    }
    name="Test2"
    onClick={[MockFunction]}
    showDeveloper={true}
    slug="test2"
    version="3.0.3-beta7483"
  />
</BaseI18n>
`;

exports[`AppTile component should render correctly an app with iconToLoad (icon being fetching) 1`] = `
<BaseI18n
  defaultLang="en"
  dictRequire={[Function]}
  lang="en"
>
  <AppTile
    app={
      Object {
        "developer": Object {
          "name": "Naming me",
        },
        "editor": "",
        "icon": "<svg></svg>",
        "iconToLoad": true,
        "installed": true,
        "installedAppLink": "test2.cozy.mock",
        "name": "Test2",
        "onClick": [MockFunction],
        "slug": "test2",
        "version": "3.0.3-beta7483",
      }
    }
    developer={
      Object {
        "name": "Naming me",
      }
    }
    editor=""
    icon="<svg></svg>"
    iconToLoad={true}
    installed={true}
    installedAppLink="test2.cozy.mock"
    name="Test2"
    onClick={[MockFunction]}
    showDeveloper={true}
    slug="test2"
    version="3.0.3-beta7483"
  />
</BaseI18n>
`;

exports[`AppTile component should render correctly an app with iconToLoad if isMobile 1`] = `
<BaseI18n
  defaultLang="en"
  dictRequire={[Function]}
  lang="en"
>
  <AppTile
    app={
      Object {
        "developer": Object {
          "name": "Naming me",
        },
        "editor": "",
        "icon": "<svg></svg>",
        "iconToLoad": true,
        "installed": true,
        "installedAppLink": "test2.cozy.mock",
        "name": "Test2",
        "onClick": [MockFunction],
        "slug": "test2",
        "version": "3.0.3-beta7483",
      }
    }
    developer={
      Object {
        "name": "Naming me",
      }
    }
    editor=""
    icon="<svg></svg>"
    iconToLoad={true}
    installed={true}
    installedAppLink="test2.cozy.mock"
    isMobile={true}
    name="Test2"
    onClick={[MockFunction]}
    showDeveloper={true}
    slug="test2"
    version="3.0.3-beta7483"
  />
</BaseI18n>
`;

exports[`AppTile component should render correctly an app without developer property 1`] = `
<BaseI18n
  defaultLang="en"
  dictRequire={[Function]}
  lang="en"
>
  <AppTile
    app={
      Object {
        "editor": "cozy",
        "icon": "<svg></svg>",
        "name": "Test",
        "namePrefix": "Cozy",
        "onClick": [MockFunction],
        "slug": "test",
        "version": "3.0.3",
      }
    }
    editor="cozy"
    icon="<svg></svg>"
    name="Test"
    namePrefix="Cozy"
    onClick={[MockFunction]}
    showDeveloper={true}
    slug="test"
    version="3.0.3"
  />
</BaseI18n>
`;

exports[`AppTile component should render correctly an installed app 1`] = `
<BaseI18n
  defaultLang="en"
  dictRequire={[Function]}
  lang="en"
>
  <AppTile
    app={
      Object {
        "developer": Object {
          "name": "Naming me",
        },
        "editor": "",
        "icon": "<svg></svg>",
        "installed": true,
        "installedAppLink": "test2.cozy.mock",
        "name": "Test2",
        "onClick": [MockFunction],
        "slug": "test2",
        "version": "3.0.3-beta7483",
      }
    }
    developer={
      Object {
        "name": "Naming me",
      }
    }
    editor=""
    icon="<svg></svg>"
    installed={true}
    installedAppLink="test2.cozy.mock"
    name="Test2"
    onClick={[MockFunction]}
    showDeveloper={true}
    slug="test2"
    version="3.0.3-beta7483"
  />
</BaseI18n>
`;

exports[`AppTile component should render correctly an installed app without icon provided 1`] = `
<BaseI18n
  defaultLang="en"
  dictRequire={[Function]}
  lang="en"
>
  <AppTile
    app={
      Object {
        "developer": Object {
          "name": "Cozy",
        },
        "editor": "cozy",
        "icon": "",
        "name": "Test",
        "namePrefix": "Cozy",
        "onClick": [MockFunction],
        "slug": "test",
        "version": "3.0.3",
      }
    }
    developer={
      Object {
        "name": "Cozy",
      }
    }
    editor="cozy"
    icon=""
    name="Test"
    namePrefix="Cozy"
    onClick={[MockFunction]}
    showDeveloper={true}
    slug="test"
    version="3.0.3"
  />
</BaseI18n>
`;
